pkg/agent/nodeportlocal/portcache/port_table.go

```diff
@@ -64,94 +93,134 @@ func NewPortTable(start, end int) (*PortTable, error) {
-	for _, data := range pt.Table {
-		if data.PodIP == ip {
-			allData = append(allData, data)
+	for i := range pt.Table {
+		if (*pt.Table[i]).PodIP == ip {
+			allData = append(allData, *pt.Table[i])
 		}
 	}
 	return allData
 }
 
-func (pt *PortTable) GetEntryByPodIPPort(ip string, port int) *NodePortData {
+func (pt *PortTable) GetEntryByPodIPPortProtocol(ip string, port int, protocol string) *NodePortData {
 	pt.tableLock.RLock()
 	defer pt.tableLock.RUnlock()
-	return pt.getEntryByPodIPPort(ip, port)
+	return pt.getEntryByPodIPPortProtocol(ip, port, protocol)
 }
 
-func (pt *PortTable) getEntryByPodIPPort(ip string, port int) *NodePortData {
-	for _, data := range pt.Table {
-		if data.PodIP == ip && data.PodPort == port {
-			return &data
-		}
+func (pt *PortTable) getEntryByPodIPPortProtocol(ip string, port int, protocol string) *NodePortData {
+	data, _ := pt.PodTable[podIPPortFormat(ip, port)]
+	if data != nil && data.HasProtocol(protocol) {
+		return data
 	}
 	return nil
 }
 
-func (pt *PortTable) getFreePort(podIP string, podPort int) (int, Closeable, error) {
+func (pt *PortTable) getEntryByPodIPPort(ip string, port int) *NodePortData {
+	pt.tableLock.RLock()
+	defer pt.tableLock.RUnlock()
+	return pt.PodTable[podIPPortFormat(ip, port)]
+}
+
+func (pt *PortTable) isNodePortAvilableForPodIPProtocol(ip string, nodeport int, podPort int, protocol string) bool {
+	val, ok := pt.Table[nodeport]
+	// A given nodeport, is marked unavailable if an entry for it already exists in the PortTable Table and
+	// the nodeport suggested is being used by another Pod or,
+	// the nodeport suggested for a Pod has a different podPort binding or,
+	// the nodeport suggested for a Pod:podPort has the protocol already configured.
+	if ok && (val.PodIP != ip ||
+		(val.PodIP == ip && val.PodPort != podPort) ||
+		(val.PodIP == ip && val.PodPort == podPort && val.HasProtocol(protocol))) {
+		return false
+	}
+	return true
+}
+
+func (pt *PortTable) getFreePort(podIP string, podPort int, protocol string) (int, Closeable, error) {
 	for i := pt.StartPort; i <= pt.EndPort; i++ {
-		if _, ok := pt.Table[i]; !ok {
-			socket, err := pt.LocalPortOpener.OpenLocalPort(i)
-			if err != nil {
-				continue
-			}
-			return i, socket, nil
+		// This ensures that the NodePort `i` is not taken by any other Pods,
+		// irrespective of the protocol.
+		if !pt.isNodePortAvilableForPodIPProtocol(podIP, i, podPort, protocol) {
+			continue
 		}
+		socket, err := pt.LocalPortOpener.OpenLocalPort(i, protocol)
+		if err != nil {
+			continue
+		}
+		return i, socket, nil
 	}
 	return 0, nil, fmt.Errorf("no free port found")
 }
 
-func (pt *PortTable) AddRule(podIP string, podPort int) (int, error) {
+func (pt *PortTable) AddRule(podIP string, podPort int, protocol string) (int, error) {
 	pt.tableLock.Lock()
 	defer pt.tableLock.Unlock()
-	nodePort, socket, err := pt.getFreePort(podIP, podPort)
+	nodePort, socket, err := pt.getFreePort(podIP, podPort, protocol)
```

Comment:
this should just be:
1. lookup the nodeport assigned for (podIP, podPort) using the new table
2. if it doesn't exist we allocate one by iterating through the old table (code is unchanged)
3. finally we call `OpenLocalPort` with the appropriate protocol